cds-snc/cppd-medical-report-sails
# MA Authentication

# What this does
- Adds basic authentication and policies for the MA routes (sessions/view/etc).
- Adds User model with checkPassword instanceMethod
- Adds [bcrypt](https://www.npmjs.com/package/bcrypt) for [password hashing](https://codahale.com/how-to-safely-store-a-password/)
- Adds isAuthenticated middleware
- Adds tests around auth functionality
- Adds command line utility for creating a user account
- Adds database seeder to create a user for testing
- Cleanup/simplify `config/policies.js`
- Adds FeatureFlags utility to enable feature flagging
- Updates Dockerfile to use multi-stage build

## To test
First, enable the Auth feature by setting `FEATURE_AUTH=true` in .env (you'll need to reload sails for the config to take effect). Then:
- attempt to access `/en/sessions`
- you should find yourself redirected to `/en/login`
- attempt to access `/en/sessions/1/view`
- you should find yourself redirected to `/en/login`
- run the seeder `npm run db:seed` to create a test user
- visit `/en/login` and use the credentials:
    - username: test@user.com
    - password: secret
- on successful login you should find yourself on the `/en/sessions` page, and have access to view sessions
- visit `/en/logout` and you should find yourself logged out and unable to access the sessions routes again
- try logging in with incorrect credentials, notice you will not get logged in

## Create user cli
Use the command line utility to create a user account:
```
sails run create-user --name="Full name" --email="name@email.com" --password="secret"
```

## Auth tests
Run the test suite by running `npm run cypress:open` and then select `Run all specs` or select the test spec to run.

Note that the Auth tests run `npm run db:seed` to ensure there is a known user for testing.

## Feature flags
This PR adds a Feature Flags utility. Feature Flags are just boolean flags specified in the environment, so to use, start by creating a feature flag in .env:

```
FEATURE_AUTH=true
```

Then, when you want to switch functionality on or off:

```
const featureFlags = require('FeatureFlags');

//

if (featureFlags.isEnabled('FEATURE_AUTH')) {
    // enable auth
}
```

## Dockerfile multi-stage build
This PR requires the `bcrypt` package for password encryption. Bcrypt needs to be built for Alpine on install, but Alpine does not have the tools to build from source. So changed our Dockerfile to a multi-stage build so we can install dependencies (python, gcc, make, etc) required for building bcrypt in an intermediate container, but then output a final image without those added dependencies.
